add cuDNN dependency and setup build for cuda4dnn

YashasSamaga opened the pull request:
### This pullrequest changes

**New files:**
- `cmake/FindCUDNN.cmake`: finds cuDNN package
  * details documented in the file
- `dnn/srcs/cuda4dnn/test.cpp`: temporary file to test compliation of files in the `cuda4dnn` directory (will be removed later)
- `dnn/srcs/cuda/test.cu`: temporary file to test compilation of files in the `cuda` directory  (will be removed later)

**New options:**
- `WITH_CUDNN` (advanced option): finds and adds cuDNN to the project
- `OPENCV_DNN_CUDA` (akin to `OPENCV_DNN_OPENCL`):
  * Requires `HAVE_CUDNN`, `HAVE_CUBLAS` and `HAVE_CUDA`
  * Adds `CUDA4DNN` as a preprocessor symbol to the DNN module sources (akin to `OCL4DNN`)

**Issues:**
- cuDNN is mostly not going to be used anywhere outside the DNN module but this PR links cuDNN to the entire project 
- haven't been able to generate the correct debian package name (check `OpenCVPackaging.cmake`)

**Tests:**
 - tested on Ubuntu 19.04 and Windows 10 with CUDA 10.1 and cuDNN 7.5

---

Reference #14585 

<details>

https://docs.nvidia.com/deeplearning/sdk/cudnn-install/index.html
Docker image: https://gitlab.com/nvidia/cuda/blob/ubuntu18.04/10.1/devel/cudnn7/Dockerfile

</details>

```
force_builders=Custom
buildworker:Custom=linux-4
docker_image:Custom=ubuntu-cuda:18.04
```

YashasSamaga commented on modules/dnn/CMakeLists.txt:
```diff
@@ -73,12 +79,19 @@ endif()
+if(OPENCV_DNN_CUDA AND HAVE_CUDA AND HAVE_CUBLAS AND HAVE_CUDNN)
+  list(APPEND include_dirs ${CUDA_TOOLKIT_INCLUDE} ${CUDNN_INCLUDE_DIRS})
+else()
+  set(sources_options EXCLUDE_CUDA)
```
[BUG] need to append `EXCLUDE_CUDA` to `sources_options`

alalek commented on cmake/templates/OpenCVConfig-CUDA.cmake.in:
```diff
@@ -14,6 +17,16 @@ else()
 
+if(OpenCV_USE_CUDNN)
+    if(NOT CUDNN_FOUND)
+        find_host_package(CUDNN ${OpenCV_CUDNN_VERSION} EXACT REQUIRED)
```
CUDNN is internal dependency (like many others).

Don't require CUDNN development files to build user applications.
See [this build](http://pullrequest.opencv.org/buildbot/builders/precommit_custom_linux/builds/2175).

YashasSamaga replied:
The cuDNN handles get exposed to the user from the DNN module includes which allows the user to write a custom layer with a CUDA implementation using cuDNN API.

I think a better way would be to abstract away cuDNN completely so that they only play with `csl::cudnn::Handle` object and `csl::Tensor` without requiring `cudnn.h`.

I'll remove this.

alalek commented:
Please fix/suppress this warning (from "Custom" builder with CUDA10.1+CUDNN):
> modules/dnn/src/cuda/test.cu(16): warning: variable "idx" was set but never used

YashasSamaga commented:
Fixed.

`CV_UNUSED` doesn't work with NVCC.

This PR was opened before the switch from `DNN_BACKEND_OPENCV` to `DNN_BACKEND_CUDA` was made. I preassume that the format for the option `OPENCV_DNN_OPENCL` is `backend_DNN_target`. Hence, I had added `OPENCV_DNN_CUDA`.

Should it be changed? `CUDA_DNN_CUDA` looks very unpleasing.

karishmathumu commented:
I am having trouble with using CUDA for the computation

I have installed opencv 4.5.6 with a CUDNN 8.6 and CUDA 11.4.
I have configured the opencv with cmake-gui, enabling, opencv_dnn_cuda and build_opencv_cuda
I use this segment in my code. 
 cout << "Using GPU device" << endl;
    net.setPreferableBackend(DNN_BACKEND_CUDA);
    net.setPreferableTarget(DNN_TARGET_CUDA);

When run, 
it switches to CPU and compiles giving the result in 400 milliseconds. 
The issue is displayed as below: 
**[ WARN:0] global /home/ubuntu/build_opencv/opencv/modules/dnn/src/dnn.cpp (1447) setUpNet DNN module was not built with CUDA backend; switching to CPU**

  Requesting for your kind help. 

Regards,
Karishma

alalek commented:
@karishmathumu Build / usage questions should go to Users OpenCV Q/A forum: https://forum.opencv.org/
